// redux-func/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.makeActionGroup = makeActionGroup;
exports.makeAjaxAction = makeAjaxAction;
exports.makeTxAction = makeTxAction;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ActionGroup = function () {
  function ActionGroup(name) {
    _classCallCheck(this, ActionGroup);

    this._name = name;
    var prefix = this._name + '.';

    for (var _len = arguments.length, ops = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      ops[_key - 1] = arguments[_key];
    }

    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = ops[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var op = _step.value;

        this[op] = prefix + op;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  _createClass(ActionGroup, [{
    key: 'clone',
    value: function clone() {
      return Object.assign(new ActionGroup(this._name), this);
    }
  }, {
    key: 'extend',
    value: function extend() {
      for (var _len2 = arguments.length, ops = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        ops[_key2] = arguments[_key2];
      }

      if (ops.length === 0) {
        return this;
      }
      var cloned = this.clone();
      var prefix = this._name + '.';
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = ops[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var op = _step2.value;

          cloned[op] = prefix + op;
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2.return) {
            _iterator2.return();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }

      return cloned;
    }
  }, {
    key: 'bind',
    value: function bind(params) {
      if (params) {
        var cloned = this.clone();
        for (var key in params) {
          if (params.hasOwnProperty(key)) {
            cloned[key] = params[key];
          }
        }
        return cloned;
      } else {
        return this;
      }
    }
  }]);

  return ActionGroup;
}();

function makeActionGroup(name) {
  for (var _len3 = arguments.length, ops = Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
    ops[_key3 - 1] = arguments[_key3];
  }

  return new (Function.prototype.bind.apply(ActionGroup, [null].concat([name], ops)))();
}

var QUERY = exports.QUERY = 'QUERY';
var SUCCESS = exports.SUCCESS = 'SUCCESS';
var ERROR = exports.ERROR = 'ERROR';
function makeAjaxAction(name) {
  for (var _len4 = arguments.length, ops = Array(_len4 > 1 ? _len4 - 1 : 0), _key4 = 1; _key4 < _len4; _key4++) {
    ops[_key4 - 1] = arguments[_key4];
  }

  return new (Function.prototype.bind.apply(ActionGroup, [null].concat([name, QUERY, SUCCESS, ERROR], ops)))();
}

var BEGIN = exports.BEGIN = 'BEGIN';
var UPDATE = exports.UPDATE = 'UPDATE';
var COMMIT = exports.COMMIT = 'COMMIT';
var CANCEL = exports.CANCEL = 'CANCEL';
function makeTxAction(name) {
  for (var _len5 = arguments.length, ops = Array(_len5 > 1 ? _len5 - 1 : 0), _key5 = 1; _key5 < _len5; _key5++) {
    ops[_key5 - 1] = arguments[_key5];
  }

  return new (Function.prototype.bind.apply(ActionGroup, [null].concat([name, BEGIN, UPDATE, COMMIT, CANCEL], ops)))();
}
